src/TrackGenerator.cpp

```diff
@@ -1554,6 +1554,67 @@ void TrackGenerator::correctFSRVolume(int fsr_id, FP_PRECISION fsr_volume) {
 
 
 /**
+ * @brief Generates the numerical centroids of the FSRs.
+ * @details This routine generates the numerical centroids of the FSRs
+ *          by weighting the average x and y values of each segment in the
+ *          FSR by the segment's length and azimuthal weight. The numerical
+ *          centroid fomula can be found in R. Ferrer et. al. "Linear Source
+ *          Approximation in CASMO 5", PHYSOR 2012.
+ */
+void TrackGenerator::generateFSRCentroids(){
+
+  int num_FSRs = _geometry->getNumFSRs();
+
+  /* Get FSR Volumes */
+  FP_PRECISION* FSR_volumes = getFSRVolumes();
+
+  /* Create temporary array of centroids and initialize to origin */
+  Point** centroids = new Point*[num_FSRs];
+  for (int r=0; r < num_FSRs; r++){
+    centroids[r] = new Point();
+    centroids[r]->setCoords(0.0, 0.0);
+  }
+
+  /* Generate the fsr centroids */
+  for (int i=0; i < _num_azim; i++) {
+    for (int j=0; j < _num_tracks[i]; j++) {
+
+      int num_segments = _tracks[i][j].getNumSegments();
+      segment* segments = _tracks[i][j].getSegments();
+      double x = _tracks[i][j].getStart()->getX();
+      double y = _tracks[i][j].getStart()->getY();
+      double phi = _tracks[i][j].getPhi();
+
+      for (int s=0; s < num_segments; s++) {
+        segment* curr_segment = &segments[s];
+        int fsr = curr_segment->_region_id;
+        double volume = FSR_volumes[fsr];
+        centroids[fsr]->setX(centroids[fsr]->getX() + _azim_weights[i] *
+                             (x + cos(phi) * curr_segment->_length / 2.0) *
+                             curr_segment->_length / FSR_volumes[fsr]);
+        centroids[fsr]->setY(centroids[fsr]->getY() + _azim_weights[i] *
+                             (y + sin(phi) * curr_segment->_length / 2.0) *
+                             curr_segment->_length / FSR_volumes[fsr]);
+
+        x += cos(phi) * curr_segment->_length;
+        y += sin(phi) * curr_segment->_length;
+      }
+    }
+  }
+
+  /* Set the centroid for the FSR */
+  for (int r=0; r < num_FSRs; r++){
+    _geometry->setFSRCentroid(r, centroids[r]);
```

Comment:
That works - I change it to that. I was just thinking it would be better for the Geometry to instantiate its own centroid in case there are changes to OpenMOC in the future where the `setFSRCentroid` method is called with a point allocated on the heap and thus destroyed with the calling method is finished.